### PR TITLE
DrivAerML: Slower cosine cycling T_max=50 at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
T_max=20 catastrophically diverged at 4L/512d (PR #2795: grad norms to 207, best 10.59% before collapse). The baseline uses T_max=30 (stable, 4.619%). On 4L/384d, T_max=50 beat T_max=30. Larger models appear to need slower cosine cycling — the high-LR phases in shorter cycles cause gradient instability in wider architectures. T_max=50 provides gentler LR transitions that should give the 4L/512d model more stability.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 50 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent rei \
  --wandb-name "rei/drivaerml-4L512d-tmax50" \
  --wandb-group "drivaerml-4L512d-tmax-sweep"
```

Report val_primary/surface_rel_l2_pct, best epoch, total epochs, and whether training remained stable throughout (no gradient explosions).

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d, T_max=30, 267 epochs, 180-min)
T_max comparison: T_max=20 diverged catastrophically. T_max=30 = baseline. T_max=50 = this test.
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT rei:

## Results

### DrivAerML: T_max=50 at 4L/512d — NEGATIVE RESULT (diverged)

| Metric | T_max=50 (this) | T_max=30 (baseline) | Delta |
|--------|-----------------|---------------------|-------|
| **val_primary/surface_rel_l2_pct** | **6.782%** (best, ep 123) | **4.619%** (best, ep 256) | +46.8% worse |
| test_primary/surface_rel_l2_pct | 47.82% (at final diverged ep) | N/A | N/A (test at wrong checkpoint) |
| Total epochs | 262 | 267 | comparable |
| Best epoch | 123 | 256 | — |
| Training stable? | **NO — two catastrophic divergences** | Stable throughout | — |
| External target | <3.71% | <3.71% | — |

### Stability Analysis

T_max=50 was **less stable** than T_max=30, not more:

1. **Epochs 1-190:** Healthy training. Grad norms 0.3-2.0. Best reached 6.782% at epoch 123.
2. **First divergence (epoch ~210):** Grad norms exploded from 2.0 → 78 → 1,717 → 3,394. Val metric collapsed from 9.8% → 65.4%.
3. **Partial recovery (epochs 224-250):** Model partially recovered to 21.6%, grad norms dropped to 15-26. But never returned to the 6-8% range.
4. **Second catastrophic divergence (epoch 253):** Grad norms exploded to 50K → 1.1M. Model completely unrecoverable at 47.6%.

### What happened

The hypothesis that T_max=50 would provide "gentler LR transitions" for 4L/512d was **falsified**. The longer cosine cycle means the LR stays near its peak for more consecutive steps. At 4L/512d scale, this extended high-LR exposure accumulates gradient instability that eventually triggers catastrophic divergence. T_max=30's shorter cycles give the model less time at peak LR, paradoxically providing more stability.

The T_max ordering for 4L/512d is now: T_max=20 (catastrophic from start) < T_max=50 (diverged at ep 210) < **T_max=30 (stable, 4.619%)**. There appears to be a sweet spot around T_max=30 where the cycle is short enough to prevent gradient accumulation but long enough for meaningful learning per cycle.

### Exact command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --agent rei \
  --wandb-name "rei/drivaerml-4L512d-tmax50" --wandb-group "drivaerml-4L512d-tmax-sweep"
```

- **W&B run:** wjjt81df
- **Peak memory:** ~97GB (RTX PRO 6000 Blackwell, single GPU)

### Suggested follow-ups

1. **Grad clipping for T_max=50:** The divergence was clearly gradient-driven. gc=1.0 or gc=1.5 with T_max=50 might prevent the explosions while preserving the deeper exploration T_max=50 provides in epochs 1-190.
2. **T_max=40 sweet spot search:** If T_max=30 is stable and T_max=50 diverges, there may be a narrow stable window around T_max=35-40 that offers better results than T_max=30.
3. **T_max=30 remains the champion for 4L/512d:** Focus scaling effort on other dimensions (grad clip, weight decay, architecture) with T_max=30 as the fixed cycling rate.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2815 because ADVISOR: Closing — your result of 6.782% does not beat the 4.619% baseline, and the T_max=50 run crashed. DrivAerML needs shorter cosine cycles, not longer. Thank you for testing this boundary, rei..
